### PR TITLE
Don't set NODELAY for non-inet sockets.

### PR DESCRIPTION
**Pull request by jdamato-fsly**
strace allows you to trace a process printing only failed system calls (which
is extremely useful for debugging). If lots of connections are generated
calling `setsockopt` on a socket that doesn't support NODELAY can fill your
strace logs with errors and make it harder to debug syscall failures you are
interested in.

This change avoids NODELAY for UNIX sockets. In addition to easier `strace`
debugging, we can also remove an unncessary system call.

### CODE REVIEW

File: lib/common/socket/evloop.c.h [hunk 1]
```diff
@@ -399,19 +399,21 @@ static struct st_h2o_evloop_socket_t *create_socket(h2o_evloop_t *loop, int fd,
...
-static struct st_h2o_evloop_socket_t *create_socket_set_nodelay(h2o_evloop_t *loop, int fd, int flags)
+static struct st_h2o_evloop_socket_t *create_socket_set_nodelay(h2o_evloop_t *loop, int fd, int flags, int sa_family)
 {
     /* ignore errors returned by setsockopt; fd is not restricted to TCP sockets (could be a unix-domain socket for example) */
-    int on = 1;
-    setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &on, sizeof(on));
+    if (sa_family != AF_UNIX) {
+        int on = 1;
+        setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &on, sizeof(on));
+    }
```
Comment (kazuho) [inline lib/common/socket/evloop.c.h:408]: I believe that this `if` is an approximation of `if (proto == IPPROTO_TCP)`?

Reducing the number of unnecessary function calls is fine, but I am concerned about the long-term impact of designing the API based on approximations.

How about removing `create_socket_set_nodelay` as a whole, and doing the following? All the comments below are based on this assumption.

File: lib/common/socket/evloop.c.h [hunk 2]
```diff
@@ -399,19 +399,21 @@ static struct st_h2o_evloop_socket_t *create_socket(h2o_evloop_t *loop, int fd,
...
     return create_socket(loop, fd, flags);
 }
 
 h2o_socket_t *h2o_evloop_socket_create(h2o_evloop_t *loop, int fd, int flags)
 {
     /* it is the reponsibility of the event loop to modify the properties of a socket for its use (i.e., set O_NONBLOCK) */
     fcntl(fd, F_SETFL, O_NONBLOCK);
-    return &create_socket_set_nodelay(loop, fd, flags)->super;
+    return &create_socket_set_nodelay(loop, fd, flags, AF_UNSPEC)->super;
```
Comment (kazuho) [inline lib/common/socket/evloop.c.h:416]: I think we can:
* call `create_socket` directly here, and
* always call `setsockopt(TCP_NODELAY)`, with a comment stating that we are too lazy to test the socket type

File: lib/common/socket/evloop.c.h [hunk 3]
```diff
@@ -447,7 +449,7 @@ h2o_socket_t *h2o_evloop_socket_accept(h2o_socket_t *_listener)
     if ((fd = cloexec_accept(listener->fd, (struct sockaddr *)peeraddr, peeraddrlen)) == -1)
         return NULL;
     fcntl(fd, F_SETFL, O_NONBLOCK);
-    sock = &create_socket_set_nodelay(listener->loop, fd, H2O_SOCKET_FLAG_IS_ACCEPTED_CONNECTION)->super;
+    sock = &create_socket_set_nodelay(listener->loop, fd, H2O_SOCKET_FLAG_IS_ACCEPTED_CONNECTION, peeraddr->sa_family)->super;
```
Comment (kazuho) [inline lib/common/socket/evloop.c.h:452]: I think we can:
* call `create_socket` directly here, and
* then call a utility function that does `if (sa_family == AF_INET || sa_family == AF_INET6) setsockopt(TCP_NODELAY)`.

File: lib/common/socket/evloop.c.h [hunk 4]
```diff
@@ -470,7 +472,7 @@ h2o_socket_t *h2o_socket_connect(h2o_loop_t *loop, struct sockaddr *addr, sockle
         return NULL;
     }
 
-    sock = create_socket_set_nodelay(loop, fd, H2O_SOCKET_FLAG_IS_CONNECTING);
+    sock = create_socket_set_nodelay(loop, fd, H2O_SOCKET_FLAG_IS_CONNECTING, addr->sa_family);
```
Comment (kazuho) [inline lib/common/socket/evloop.c.h:475]: Same as above.

**Review by kazuho — COMMENTED** [review]
Thank you for the PR. I agree that it is good to reduce the number of unnecessary syscalls. OTOH I have some concern regarding the approach, and am suggesting an alternative below. Please let me know what you think.

**Comment by jdamato-fsly** [comment]
OK, I've updated this PR with your suggestions, @kazuho. Thank you for taking a look!

**Review by kazuho — APPROVED** [review]
Thank you for the changes. LGTM. To me be merged with small editorial modifications.

**Pull request merged**